[PROTOCOL RFC] Catalog-managed Tables (#4382)
delta-io/delta: protocol_rfcs/catalog-owned.md

accesses tables.

Today’s Delta protocol relies entirely on the filesystem for read-time discovery as well as
write-time commit atomicity. This feature request is to allow catalog-owned Delta tables whose
discovery and commits go through the table's owning catalog instead of going directly to the
filesystem (s3, abfs, etc). In particular, the catalog becomes the source of truth about whether a
given commit attempt succeeded or not, instead of relying exclusively on filesystem PUT-if-absent
primitives.

Making the catalog the source of truth for commits to a table brings several important advantages:

1. Allows the catalog to broker all commits to the tables it owns, and to reject filesystem-based
   commits that would bypass the catalog. Otherwise, the catalog cannot reliably stay in sync with the
   table state, nor can it reject invalid commits, because it doesn’t even know about writes until they
   are already durable and visible to readers. For example, a catalog would want to block attempts to
   drop the NOT NULL constraint on a column which is referenced by a FOREIGN KEY constraint in a

>    different table.

Author: You mean in our Delta Catalog? Hm. I'm not sure. I'd think this would instead be the responsibility of, say, a Glue catalog or Unity catalog implementation.

cc @scovich, @LukasRupprecht

>    different table.

Reviewer: > Is this functionality already implemented in the current Delta codebase? For Catalog-Owned enabled tables, should Catalog Client takes the responsibility to check/block the attempt - or it depends on the actual Catalog implementation?

This sort of functionality belongs to the catalog's own client (Glue, UC, etc) that interacts with Delta. The RFC only cares to ensure that catalogs are _able_ to perform such checks, not whether/how they do it.

   `_delta_log/_staged_commits/<v>.<uuid>.json`.
    - Here, the `uuid` is a random UUID that is generated for each commit and `v` is the version
      which is being committed, zero-padded to 20 digits.
    - The catalog tracks only the location, not the content, of a staged commit.
    - The mere existence of one of these files does not mean that the file is a _ratified_ commit.
      It might correspond to a failed or in-progress commit attempt.
    - The catalog is the source of truth around which staged commits are "ratified".

3. **Inline commit** <a name="inline-commit"></a>: A proposed commit that is not written to disk but
   rather has its content sent to the catalog for the catalog to track directly.

4. **Ratified commit**: A proposed commit that a catalog has determined has won the commit at the
   desired version of the table.
    - It may or may not yet be "published".
    - It may or may not even be _tracked_ by the catalog at all - the catalog may have just
      atomically published it to the filesystem directly, relying on PUT-if-absent primitives to
      facilitate the ratification and publication all in one step.

>     - The catalog must track ratified commits until they are published to the `_delta_log` directory.

Author: Great feedback, thanks!

must LIST the `_delta_log` directory to get information about the published commits. For commits
that are both returned by the catalog and already published, Delta clients can choose to read
either. Delta clients must ignore any files with versions greater than the latest ratified commit
version returned by the catalog.

Combining these two sets of files and commits enables Delta clients to generate a snapshot at the
latest version of the table.

**NOTE**: This spec prescribes the _minimum_ required interactions between Delta clients and
catalogs for commits. Catalogs may very well expose APIs and work with Delta clients to be
informed of other non-commit [file types](#file-types), such as checkpoint, log
compaction, and version checksum files. This would allow catalogs to return additional
information to Delta clients during query and scan planning, potentially allowing Delta
clients to avoid LISTing the filesystem altogether.

## Commit Protocol

To start, Delta Clients send the desired actions to be commited to the client-side component of the
catalog.

This component then has several options for proposing, ratifying, and publishing the commit,
detailed below.

> - Write the actions (likely client-side) to an [staged commit file](#staged-commit) in the

Author: This makes it much clearer, thanks!

  `_delta_log/_staged_commits` directory and then ratify the staged commit by atomically recording
  (likely server-side) (in persistent storage of some kind) that the file corresponds to version
  `v`.

> - Treat this as an [inline commit](#inline-commit) (i.e. likely that the client-side component sends

Author: Yup!

  the contents to the server-side component) and atomically record the content of the commit as
  version `v` of the table.
- Propose, ratify, and publish all-in-one by atomically writing a [published commit file](#published-commit)
  in the `_delta_log` directory. Note that this commit will be considered to have succeeded as soon
  as the file becomes visible in the filesystem, regardless of when or whether the catalog is made
  aware of the successful publish. The catalog does not need to track these files.

Note that when a catalog receives a request to ratify version `v`, it must first verify that the
previous version `v-1` already exists (ratified or published), and that version `v` does not yet

> exist.

Author: I like this 👍

     *
     * @return CommitResponse which has details around the new committed delta file.
     */
    def commit(
        version: Long,
        actions: Iterator[String]): CommitResponse

    /**
     * Retrieves a (possibly empty) suffix of ratified commits in the range [startVersion,
     * endVersion] for this table.
     * 
     * Some of these ratified commits may already have been published. Some of them may be staged,
     * in which case the staged commit file path is returned, while others may be inline, in which
     * case the inline commit content is returned.
     * 
     * The returned commits are sorted in ascending version number. The returned commits may not be
     * contiguous.
     *
     * If neither start nor end version is specified, the catalog will return all available ratified
     * commits (possibly empty, if all commits have been published).
     *
     * In all cases, the response also includes the table's latest ratified commit version.
     *
     * @return GetCommitsResponse which contains an ordered list of ratified commits
     *         tracked by the catalog, as well as table's latest commit version.
     */

>     def getRatifiedCommits(

Author: Should this be `getTable` instead?

1. Allows the catalog to broker all commits to the tables it owns, and to reject filesystem-based
   commits that would bypass the catalog. Otherwise, the catalog cannot reliably stay in sync with the
   table state, nor can it reject invalid commits, because it doesn’t even know about writes until they
   are already durable and visible to readers. For example, a catalog would want to block attempts to
   drop the NOT NULL constraint on a column which is referenced by a FOREIGN KEY constraint in a
   different table.

2. Opens a clear path to transactions that could span multiple tables and/or involve non-table
   catalog updates. Otherwise, the catalog cannot participate in commit at all, because
   filesystem-based commits (i.e. using PUT-if-absent) do not admit any way to coordinate with other
   entities.

3. Allows the catalog to facilitate efficient writes of the table, e.g. by directly hosting the
   content of small commits instead of forcing clients to write them to cloud storage first. Otherwise,
   the catalog is not a source of truth, and at best it can only mirror stale copies of table state.

> 4. Allows the catalog to facilitate efficient reads of the table. Examples include vending storage

Reviewer 2: In theory the caching of such files doesn't need to be done by the catalog because they are immutable.

What is a great advantage of a catalog is that you can avoid having to do a LIST to get the latest committed version. That can save time, and it can also allow implementations of Delta on file systems where LIST is not ordered, such as S3 Express One Zone.

`00000000000000000100.<uuid>.json` always corresponds to a commit attempt for version 100. Besides
simplifying implementations, it also acknowledges the fact that commit files cannot safely be reused
for multiple commit attempts. For example, resolving conflicts in a table with [row
tracking](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#row-tracking) enabled requires
rewriting all file actions to update their `baseRowId` field.</ins>

<ins>The [catalog](#terminology-catalogs) is the source of truth about which staged commit files in
the `_delta_log/_staged_commits` directory correspond to ratified versions, and Delta clients should
not attempt to directly interpret the contents of that directory. Refer to
[catalog-owned tables](#catalog-owned-tables) for more details.</ins>

~~Delta files use new-line delimited JSON format, where every action is stored as a single line JSON
document. A delta file, `n.json`, contains an atomic set of [_actions_](#actions) that should be
applied to the previous table state, `n-1.json`, in order to the construct `n`th snapshot of the
table. An action changes one aspect of the table's state, for example, adding or removing a file.~~

> <ins>Delta files use newline-delimited JSON format, where every action is stored as a single-line

Reviewer 2: This uses "Delta files", but the preceding text uses "commits". And the section title says "Delta Log Entries". Nothing actually says that a Delta Log Entry == a Delta file == a commit.

It might make sense to move this paragraph above the whole catalogOwned sequence. The first part is really an introduction into what's in these Delta files, and then the catalogOwned is a modification to that. The intermezzo about catalogOwned is really making it hard to pop back up the stack to this higher level intro description.

`v-1`, in order to construct the `v`th snapshot of the table. An action changes one aspect of the
table's state, for example, adding or removing a file.</ins>

### Commit Provenance Information

> ***Change to [existing section](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#commit-provenance-information)***

<ins>When the `catalogOwned` table feature is enabled, the `commitInfo` action must have a field
`txnId` that stores a unique transaction identifier string.</ins>

### Metadata Cleanup

> ***Change to [existing section](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#metadata-cleanup)***

2. Identify the newest checkpoint that is not newer than the `cutOffCommit`. A checkpoint at the
   `cutOffCommit` is ideal, but an older one will do. Lets call it `cutOffCheckpoint`. We need to
   preserve the `cutOffCheckpoint` and all <ins>published</ins> commits after it, because we need
   them to enable time travel for commits between `cutOffCheckpoint` and the next available
   checkpoint.
    - <ins>If no `cutOffCheckpoint` can be found, do not proceed with metadata cleanup as there is
      nothing to cleanup.</ins>

> 3. Delete all [delta log entries](#delta-log-entries) and [checkpoint files](#checkpoints) before

Reviewer 2: Existing: should we have the CRC files listed here as well for deletion?

Author: Yes!

The table feature defines the parts of the [commit protocol](#commit-protocol) that directly impact
the Delta table (e.g. atomicity requirements, publishing, etc). The Delta client and catalog
together are responsible for implementing the Delta-specific aspects of commit as defined by this
spec, but are otherwise free to define their own APIs and protocols for communication with each
other.

**NOTE**: Filesytem-based access to catalog-owned tables is not supported. Delta clients are
expected to discover and access catalog-owned tables through the owning catalog, not by direct
listing in the filesystem. This feature is primarily designed to warn filesystem-based readers that
might attempt to access a catalog-owned table's storage location without going through the catalog
first, and to block filesystem-based writers who could otherwise corrupt both the table and the
catalog by failing to commit through the catalog.

Before we can go into details of this protocol feature, we must first align our terminology.

## Terminology: Commits

A commit is a set of [actions](#actions) that transform a Delta table from version `v - 1` to `v`.

> A commit contains the same content as a [Delta file](#delta-log-entries).

Reviewer 2: We could say "A commit may be stored in the file system as a Delta file." A commit doesn't always contain the content of a (particular) Delta file, i.e., there may not actually be a Delta file for a commit.

Author: > A commit doesn't always contain the content of a (particular) Delta file, i.e., there may not actually be a Delta file for a commit.

Hi Bart. The goal here was to communicate the general rules of the content that commits contain. The goal isn't to communicate where or how a commit can be stored (e.g. as you said, "in the file system as a delta file".

So --> I don't quite understand your emphasis of it always containing the same content as a _particular_ Delta file? Could you please elaborate?

I am certainly traying to convey that: commits represent the same information, thus "contain the same content" as Delta Log Entries. They contain a list of actions.

I realize now, by the way, that inline commits, for example, do *not* have to store actions as newline-delimited JSON? inline commits can probably be stored in whatever format the catalog wants?

Reviewer 2: > So --> I don't quite understand your emphasis of it always containing the same content as a particular Delta file? Could you please elaborate?
It's the way it is phrased. "A commit contains the same content as a Delta file" can be interpreted as "it contains the same content as a particular Delta file, and a commit only exists by the grace of the existence of that Delta file". It could also be interpreted as "It contains the same *kind of* content as a Delta file", which is the interpretation that you mean. I was proposing to clarify this ambiguity.

> I realize now, by the way, that inline commits, for example, do not have to store actions as newline-delimited JSON? inline commits can probably be stored in whatever format the catalog wants?
Agreed!

Author: > It's the way it is phrased. "A commit contains the same content as a Delta file" can be interpreted as "it contains the same content as a particular Delta file, and a commit only exists by the grace of the existence of that Delta file". It could also be interpreted as "It contains the same kind of content as a Delta file", which is the interpretation that you mean. I was proposing to clarify this ambiguity.

Ah, thanks for explaining! Makes sense.

There are several types of commits:

> 1. **Proposed commit**:  A commit that a Delta client has proposed for the next version of the

Reviewer 2: "Has proposed" suggests that the client must have already contacted the catalog. A staged commit is a proposed commit, but what if I write a file in _delta_log/_staged_commits but wait with actually proposing it to the catalog? Is it still a proposed commit if I haven't proposed it? Can it be a staged commit if it's not a proposed commit?

We could work around this issue by calling this a "prepared commit". Preparing doesn't require that you already contact the catalog.

Author: > [From the Commit Protocol section] To start, Delta Clients send the desired actions to be commited to the client-side component of the
catalog.
This component then has several options for proposing, ratifying, and publishing the commit,
detailed below.

I am totally open to that suggestion Bart. And thanks for calling out those great questions. My worry with "prepared" is that it seems you jump from "prepared" to "ratified" without any "proposing"? And I am inclined to avoid including _both_ prepared and proposed?

I think this section should actually say `A commit that the _client-side catalog component_ has proposed for ...`

- Step 1: Delta Clients send actions to the client-side catalog component
- Step 2: Client-side catalog component must **propose** them. It can write them to a staged commit file or send them inline to the server-side catalog component
- Suppose the client-side component wrote to a staged commit file but then never forwarded that location to the server-side catalog component. I'm okay still calling that a proposed commit? It's a proposed commit that the server-side catalog component never learned about, but the Delta Client sent the actions and the client-side catalog component did _something_ with them. It never got ratified -- this all seems fine?

I'm not trying to push back _too_ hard against your suggestion btw. Let me know what you think of the above. Thanks!

Reviewer 2: "Prepared" really is "staged", right? I.e., it lives in the `staged_commits` directory. I think the key problem that I have with the phrasing isn't here, but it is in the definition of "staged commit", where it says that a staged commit is always a proposed commit. What I'm saying is that a staged commit doesn't have to be proposed! The text below actually says that it could be a failed commit attempt. But I think it could also be a commit attempt that hasn't even reached the "proposed" state yet.

Author: > What I'm saying is that a staged commit doesn't have to be proposed!

Sounds good 👍

   table. It could be "staged" or "inline". It will either become "ratified" or be rejected.

> 2. **Staged commit** <a name="staged-commit"></a>: A proposed commit that is written to disk at

Reviewer 2: Do we need to say that this has the format of a normal Delta file?

   `_delta_log/_staged_commits/<v>.<uuid>.json`.
    - Here, the `uuid` is a random UUID that is generated for each commit and `v` is the version

>       which is being committed, zero-padded to 20 digits.

Reviewer 2: ```suggestion
      which is proposed to be committed, zero-padded to 20 digits.
```

    - The catalog tracks only the location, not the content, of a staged commit.
    - The mere existence of one of these files does not mean that the file is a _ratified_ commit.
      It might correspond to a failed or in-progress commit attempt.
    - The catalog is the source of truth around which staged commits are "ratified".

> 3. **Inline commit** <a name="inline-commit"></a>: A proposed commit that is not written to disk but

Reviewer 2: Inline commits can also be ratified, right?

Author: Yes. I think I need to make it clearer that both staged commits and inline commits are (1) proposed and then (2) ratified.

   rather has its content sent to the catalog for the catalog to track directly.

4. **Ratified commit**: A proposed commit that a catalog has determined has won the commit at the
   desired version of the table.
    - The catalog must track ratified commits until they are published to the `_delta_log` directory.
    - A ratified commit may or may not yet be published.

>     - A ratified commit may or may not even be _tracked_ by the catalog at all - the catalog may

Reviewer 2: The concept of "tracked" is new. We could say that the state for unpublished ratified commits is always stored by the catalog, while published commits may or may not be stored by the catalog. ("Stored" makes it explicit what "tracked" means.)

Changing this phrasing would require changes across the RFC. I could also live with phrasing where the catalog "keeps track of" commits. The use of "to track" doesn't work for me. To me (and to the dictionary definition), tracking means following the direction or location of something that moves or changes, but we use it just to mean "store".

Author: I'd be happy to change this to "store" -- why don't we do that?

      have just atomically published it to the filesystem directly, relying on PUT-if-absent
      primitives to facilitate the ratification and publication all in one step.

5. **Published commit** <a name="published-commit"></a>: A ratified commit that has been copied into
   the `_delta_log` as a normal Delta file, i.e. `_delta_log/<v>.json`.
    - Here, the `v` is the version which is being committed, zero-padded to 20 digits.
    - The existence of a `<v>.json` file proves that the corresponding version `v` is ratified,
      regardless of whether the table is catalog-owned for filesystem-based. The catalog is allowed
      to return information about published commits, but Delta clients can also use filesystem
      listing operations to directly discover them.
    - Published commits do not need to be tracked by the catalog.

## Terminology: Delta Client

This is the component that implements support for reading and writing Delta tables, and implements
the logic required by the `catalogOwned` table feature. Among other things, it
- triggers the filesystem listing, if needed, to discover published commits
- generates the commit content (the set of [actions](#actions))

> - works together with the query engine to trigger the commit process and invoke the client-side

Reviewer 3: `Delta Client` is also responsible for **defining** the interface/APIs/contract with the client-side catalog component (i.e., `Catalog Client`), right?

Author: Yes. Added!

  catalog component with the commit content

## Terminology: Catalogs

1. **Catalog**: A catalog is a entity which manages a Delta table, including its creation, writes,
   reads, and eventual deletion.
    - It could be backed by a database, a filesystem, or any other persistence mechanism.

>     - Each catalog has its own spec around how Delta clients should contact them, and how they

Reviewer 2: "contact them" implies the presence of a server. How about "interact with them"?

      perform a commit.

2. **Catalog Client**: The catalog always has a client-side component which the Delta client
   interacts with directly.
    - The Delta client is responsible for defining the client-side API that catalogs should target.
    - This client-side component has two primary responsibilities:
        - implement any client-side catalog-specific logic (such as [publishing](#publishing-commits))
        - communicate with the Catalog Server, if any

3. **Catalog Server**: The catalog may also involve a server-side component which the client-side
   component would be responsible to communicate with.
    - This server is responsible for coordinating commits and potentially persisting table metadata
      and enforcing authorization policies.

>     - Not all catalogs require a server; some may be entirely client-side (e.g. filesystem-backed
>       catalogs).

Reviewer 2: ```suggestion
    - Not all catalogs require a server; some may be entirely client-side, e.g. filesystem-backed
      catalogs, or they may make use of a generic database server and implement all of the logic
      client-side.
```

## Catalog Responsibilities

When the `catalogOwned` table feature is enabled, a catalog performs commits to the table on behalf
of the Delta client.

As stated above, the Delta spec does not mandate any particular client-server design or API for
catalogs that manage Delta tables. However, the catalog does need to provide certain capabilities
for reading and writing Delta tables:

- Atomically commit a version `v` with a given set of `actions`. This is explained in detail in the
  [commit protocol](#commit-protocol) section.
- Retrieve information about recent ratified commits and the latest ratified version on the table.
  This is explained in detail in the [Getting Ratified Commits from the Catalog](#getting-ratified-commits-from-the-catalog) section.
- Though not required, it is encouraged that catalogs also return the latest table-level metadata,
  such as the latest Protocol and Metadata actions, for the table. This can provide significant
  performance advantages to conforming Delta clients, who may forgo log replay and instead trust
  the information provided by the catalog during query planning.

## Reading Catalog-owned Tables

> A catalog-owned table can have a mix of (a) published and (b) ratified and non-published commits.

Reviewer 2: ```suggestion
A catalog-owned table can have a mix of (a) published and (b) ratified but non-published commits.
```

either. Delta clients must ignore any files with versions greater than the latest ratified commit
version returned by the catalog.

Combining these two sets of files and commits enables Delta clients to generate a snapshot at the
latest version of the table.

**NOTE**: This spec prescribes the _minimum_ required interactions between Delta clients and
catalogs for commits. Catalogs may very well expose APIs and work with Delta clients to be
informed of other non-commit [file types](#file-types), such as checkpoint, log
compaction, and version checksum files. This would allow catalogs to return additional
information to Delta clients during query and scan planning, potentially allowing Delta
clients to avoid LISTing the filesystem altogether.

## Commit Protocol

To start, Delta Clients send the desired actions to be commited to the client-side component of the
catalog.

This component then has several options for proposing, ratifying, and publishing the commit,
detailed below.

- Option 1: Write the actions (likely client-side) to an [staged commit file](#staged-commit) in the
  `_delta_log/_staged_commits` directory and then ratify the staged commit by atomically recording

>   (likely server-side) (in persistent storage of some kind) that the file corresponds to version

Reviewer 2: Double parenthesized asides looks weird :)

Author: Fixed. @bart-samwel -- does the catalog _have_ to store the location of the staged ratified commit and the content of the inline ratified commit in persistent storage?

Can't it keep track of them purely in memory?

Reviewer 2: If the catalog process dies, it shouldn't lose the information on ratified but unpublished commits. In theory it could store things in memory, if that memory is replicated N ways across multiple regions and has a lot of 9s of reliability. But at that point it's actually "persistent" in spirit. :)

  `v`.
- Option 2: Treat this as an [inline commit](#inline-commit) (i.e. likely that the client-side
  component sends the contents to the server-side component) and atomically record the content of
  the commit as version `v` of the table.
- Option 3: Propose, ratify, and publish all-in-one by atomically writing a [published commit file](#published-commit)
  in the `_delta_log` directory. Note that this commit will be considered to have succeeded as soon
  as the file becomes visible in the filesystem, regardless of when or whether the catalog is made

>   aware of the successful publish. The catalog does not need to track these files.

Reviewer 2: This paragraph makes it sound like the catalog client can always publish a commit directly without contacting the catalog server. The last sentence doesn't help, it just reinforces that "the catalog (server) is not involved at all". I think this particular case needs to be strongly qualified that this approach only works for catalog clients that use PUT-if-absent. We can also mention the option that the server does the PUT-if-absent.

A catalog must not ratify version `v` until it has ratified version `v - 1`, and it must ratify
version `v` at most once.

The catalog must track both flavors of ratified commits (staged or inline) and make them available
to readers until they are [published](#publishing-commits).

> Delta clients are encouraged to establish an API contract where the catalog provides the latest

Reviewer 2: Mention "for performance reasons"? Because it's really just a performance optimization.

[published](#publishing-commits).

The catalog is responsible to implement an API (defined by the Delta client) that Delta clients can
use to retrieve the latest ratified commit version (authoritative), as well as the set of ratified
commits the catalog is still tracking for the table. If some commits needed to complete the snapshot
are not tracked by the catalog, as they are already published, Delta clients can issue a filesystem
LIST operation to retrieve them.

Delta clients must establish an API contract where the catalog provides ratified commit information
as part of the standard table resolution process performed at query planning time.

## Publishing Commits

Publishing is the process of copying the ratified commit with version `<v>` to
`_delta_log/<v>.json`. The ratified commit may be a staged commit located in
`_delta_log/_staged_commits/<v>.<uuid>.json`, or it may be an inline commit whose content the
catalog tracks itself. Because the content of a ratified commit is immutable, it does not matter
whether the client-side, server-side, or both catalog components initiate publishing.

> Implementations are strongly encouraged to publish commits promptly. This reduces the number of

Reviewer 2: This is not necessary. In theory a catalog could be fine tracking many commits, and clients could benefit from getting those commits from the catalog instead of from the file system. E.g. if checkpoint interval is 2500 commits, then it might be beneficial if the catalog serves up 2500 commits.

I would suggest simply making the choice of when to publish the joint responsibility of the client and the catalog.

Author: I'll just delete this sentence then?

the need for synchronization or coordination between conurrent writers.

Because of this, "gaps" may appears in the `_delta_log` in the filesystem: cases where versions `v`
and `v + 2` are published, but version `v + 1` is not.

**NOTE**: Because commit publishing can happen at any time and in any order after the commit
succeeds, the file modification timestamp of the published file will not accurately reflect the
original commit time. For this reason, catalog-owned tables must use
[in-commit-timestamps](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#in-commit-timestamps)
to ensure stability of time travel reads. Refer to
[Writer Requirements for Catalog-Owned Tables](#writer-requirements-for-catalog-owned-tables)
section for more details.

## Maintenance Operations on Catalog-owned Tables

Delta clients must follow the owning catalog's rules, if any, when performing maintenance
operations like [Checkpoints](#checkpoints-1), [Log Compaction Files](#log-compaction-files),
[Metadata Cleanup](#metadata-cleanup), and VACUUM. These rules can be unique to each catalog and may
evolve over time.

> In particular, maintenance operations like Checkpoints, Log Compaction Files, and Metadata Cleanup

Reviewer 2: The previous paragraph says that all of these things are based on the catalog's rules, but then this says "in particular", as if it's an example of a catalog-based rule, but it actually states a universal rule, not a rule that a catalog may choose to require or not.

can only be performed on published versions, unless the catalog specifies otherwise.

> To highlight: Because the catalog controls the lifetime and potentially the storage of the table,

Reviewer 2: This is also not a catalog rule, it's a universal rule. If the catalog is OK with the clients vacuuming, then that's OK. Clients should not try to VACUUM the table without having the catalog's permission (according to whatever the catalog's rules are).

Delta clients would primarily be involved with the second step, but an implementation could choose
to combine the second and third steps so that a single catalog call registers the table as part of
the table's first commit.

As another example, dropping a table can be as simple as removing its name from the catalog (a "soft
delete"), followed at some later point by a "hard delete" that physically purges the data. The Delta
client would not be involved at all in this process, because no commits are made to the table.

## Catalog-owned Table Enablement

The `catalogOwned` table feature is supported and active when:
- The table is on Reader Version 3 and Writer Version 7.
- The table has a `protocol` action with `readerFeatures` and `writerFeatures` both containing the
  feature `catalogOwned`.

## Writer Requirements for Catalog-owned tables

When supported and active:

- Writers must discover and access the table using catalog calls, which happens before the table's
  protocol is known.
- The [in-commit-timestamps](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#in-commit-timestamps)
  table feature must be supported and active.

> - The `commitInfo` action must also a field `txnId` that stores a unique transaction identifier

Reviewer 2: ```suggestion
- The `commitInfo` action must also contain a field `txnId` that stores a unique transaction identifier
```

  string
- Writers must follow the catalog's [commit protocol](#commit-protocol) and must not perform
  ordinary filesystem-based commits against the table.
- Writers must follow the catalog's [maintenance operation protocol](#maintenance-operations-on-catalog-owned-tables)

## Reader Requirements for Catalog-owned tables

When supported and active:

- Readers must discover the table using catalog calls, which happens before the table's protocol
  is known.
- Readers must contact the catalog for information about unpublished ratified commits.

> ## Sample client API

Reviewer 2: ```suggestion
## Sample Catalog Client API
```

     */
    def commit(
        version: Long,
        actions: Iterator[String]): CommitResponse

    /**
     * Retrieves a (possibly empty) suffix of ratified commits in the range [startVersion,
     * endVersion] for this table.
     * 
     * Some of these ratified commits may already have been published. Some of them may be staged,
     * in which case the staged commit file path is returned; others may be inline, in which case
     * the inline commit content is returned.
     * 
     * The returned commits are sorted in ascending version number. The returned commits may not be
     * contiguous.
     *
     * If neither start nor end version is specified, the catalog will return all available ratified
     * commits (possibly empty, if all commits have been published).
     *
     * In all cases, the response also includes the table's latest ratified commit version.
     *
     * @return GetCommitsResponse which contains an ordered list of ratified commits
     *         tracked by the catalog, as well as table's latest commit version.
     */
    def getRatifiedCommits(
        startVersion: Option[Long],
        endVersion: Option[Long]): GetCommitsResponse
}
```

> Note that the above is only one example of a possible Delta Client API. It is also _NOT_ a catalog

Reviewer 2: ```suggestion
Note that the above is only one example of a possible Catalog Client API. It is also _NOT_ a catalog
```
